### ISSUE TITLE
Received fatal alert: handshake_failure

### ISSUE THREAD

**Issue by kwladyka** [http-kit/http-kit#430]
```
@(http/request
     {:timeout 1000
      :url "https://api.redjar.net"})
```

```
at java.base/sun.security.ssl.SSLEngineImpl.beginHandshake(SSLEngineImpl.java:102)
Received fatal alert: handshake_failure
```

The URL works with `curl` or `clj-http`. I also checked SSL with tools online and this is fine. The issue is only with http-kit.

I was trying with both
[http-kit "2.3.0"]
[http-kit "2.4.0-alpha6"]

---
Edit:
I guess this is about `SNI`. I see this is not documented, but there is some kind of way to add this to http-kit client.

But can it work by default? I think it should.

**Comment by ptaoussanis** [comment]
Please see [here](https://github.com/http-kit/http-kit#enabling-http-kit-client-sni-support), good luck!

**Comment by kwladyka** [comment]
Oh I didn't notice this here. It can save you from situation like this if add this to the doc here http://http-kit.github.io/client.html which was the place where I was looking solution.

**Comment by lwhorton** [comment]
For those still a little confused (the documentation link moved to [here](https://github.com/http-kit/http-kit/wiki/Client-docs#server-name-indication-sni), the problem is that http-kit clients <2.7 don't have SNI enabled by default.

**Comment by GAumala** [comment]
SNI is enabled correctly on 2.7.0, but not on 2.8.0.  I had to downgrade to fix the `handshake_failure` issue.

**Comment by ptaoussanis** [comment]
> @GAumala SNI is enabled correctly on 2.7.0, but not on 2.8.0. I had to downgrade to fix the handshake_failure issue.

Hi Gabriel, can you please provide some context for this statement? What makes you believe that SNI is not enabled correctly on 2.8.0?

**Comment by GAumala** [comment]
@ptaoussanis Last night I was working on a web app that uses http-kit's client to call some public APIs. I was adding a new API that presented the `handshake_failure` error. After finding this thread I upgraded from `2.4.0` to `2.8.0`, but it didn't fix the issue. Only after downgrading to `2.7.0` it was finally fixed.

This morning I tried to reproduce the issue, but had no luck. There was probably another issue with my code last night. Today I upgraded to 2.8.0 in my original project and it works fine. I even tested it on a VPS in NY and it also works fine.

Sorry for the confusion!

**Comment by borkdude** [comment]
Could perhaps be an AOT issue with leftovers from an older version

https://www.michielborkent.nl
https://www.eetvoorjeleven.nu

**Comment by ptaoussanis** [comment]
@GAumala No worries Gabriel, thanks for the update. Could be an AOT issue as Michiel mentioned, or other random gremlins :-)

**Comment by gabrielgiussi** [comment]
Hi @ptaoussanis, sorry to bring back this issue.
I'm currently using http-kit `2.5.2` and configuring SNI with this
```clojure
(defn ^:private sni-configure [^SSLEngine ssl-engine, ^URI uri]
  (let [^SSLParameters ssl-params (.getSSLParameters ssl-engine)]
    (.setServerNames ssl-params [(SNIHostName. (.getHost uri))])
    (.setSSLParameters ssl-engine ssl-params)))

(defn ^:private make-client
  []
  (http-kit/make-client {:ssl-configurer sni-configure}))
```

But bumping to `2.8.0` starts failing requests with
> "cause": "Received fatal alert: handshake_failure"

Should I change the way to configure SNI if I bump to 2.8.0?

**Comment by ptaoussanis** [comment]
@gabrielgiussi Hi Gabriel,

v2.8.0 (actually v2.7.0+) should use the SNI client by default, so you should be able to remove your SNI configuration code.

> Note: the request is using insecure? true

This could potentially be an issue (see [#528](https://github.com/http-kit/http-kit/issues/528)) for context, you might need to test to make sure.

Hope that's at least some help!
